### DISCUSSION TITLE
Hooks / On Delete

### DISCUSSION THREAD

**Discussion by kamalkech** [typegoose/typegoose#650] [Q&A]
Using hook when try to delete a doc, how can i get this object of doc ??

```typescript
@pre('deleteOne', async function (next) {
  const _id = this.getFilter()._id;
  // Todo: need get full object that want to delete !!
  next();
})
```
and if not accept object param ```{ document: false, query: true }``` so code bellow not working !!
```typescript
@pre('deleteOne', { document: false, query: true }, async function (next) {
  const _id = this.getFilter()._id;
  // Todo: need get full object that want to delete !!
  next();
})
```

any idea to fixe this issue ?

**Comment by smolinari** [comment]
It seems you are calling the `.deleteOne` via a query and not via a model. Try doing a `.deleteOne` via the model. That should get you `this` as the document to be deleted, I believe. 

Scott

**Reply by hasezoey** [reply]
also, in addition the `this` parameter may need to be manually set or the generic on `pre`/`post` to be set

also, mongoose is weird, the document may also be (in some cases) be a argument next to `next`

**Comment by kamalkech** [comment]
@smolinari to call .deleteOne i am using a model on the service :

```typescript
@Injectable()
export class ArticleService {
  constructor(
    @InjectModel(ArticleModel)
    private readonly articleModel: ReturnModelType<typeof ArticleModel>
  )

  delete = async (_id: mongoose.Types.ObjectId): Promise<any> => {
    try {
      return await this.articleModel
        .deleteOne({ _id })
        .exec();
    } catch (e) {
      throw e;
    }
  };
}
```

**Reply by hasezoey** [reply]
in a github discussion, please use the `Write a Reply`(to the answer (in this case from @smolinari )) instead of `Suggest an Answer`(writing an answer to your original question)

---

> and if not accept object param { document: false, query: true } so code bellow not working !!

please look at the [documentation](https://typegoose.github.io/typegoose/docs/api/decorators/hooks#pre) / typescript types first, because hook option in typegoose are the *third* parameter, not the *second*

**Reply by kamalkech** [reply]
@hasezoey still not get the doc
```typescript
@pre(
  'deleteOne',
  async function (next) {
    console.log('this :>> ', this);
  },
  { document: true, query: true },
)
```

**Reply by hasezoey** [reply]
maybe its an argument to the function?

you can test with:
```ts
async function (...args): {
  console.log("arguments", ...args);
  // or
  console.log("arguments, args);
}
```

note: this hook (`deleteOne`) does not have typeings in typegoose, because i was not sure what the correct typeings were, and since then were not really updated (type wise)

**Reply by kamalkech** [reply]
@hasezoey i am sorry if i said your last code not working, the result is args = [] or empty

**Reply by hasezoey** [reply]
i just tested the hook, and everything seems to be working fine for me, though the hook for `deleteOne` is a Query hook and can be accessed with `this`

(Note: in a "Query" pre-hook, you dont have access to any document, only the query before being executed, so you could modify query options or how documents are searched - but you wont have access to any document properties, from what i know)

reproduction repository branch: [typegoose-testing: verify650](https://github.com/typegoose/typegoose-testing/tree/verify650)
